feat: Add crossbuild for AARCH64 on X86_64 for relay-cabi lib

olksdr opened the pull request:
These changes basically add the possibility to cross compile `relay-cabi` library with proper environment, correct version of libc (the one which is in `manylinux2014_aarch64` - **2.17**) and other required dependencies.

The main challenge is to assemble the correct env, old enough to have very old libc version and the required libs. That's why as the base for building this env `centos:7` image was chosen. The newer versions of the images (e.g. recent `ubuntu`) do not have such challenges, or we could just use [`cross`](https://github.com/cross-rs/cross) which is included in [`action-rs/cargo`](https://github.com/actions-rs/cargo#cross-compilation) which we already use.


 
-----


###  Test builds comparison:

#### New release build

<img width="521" alt="image" src="https://user-images.githubusercontent.com/1931331/186834239-abf6a6dd-c01f-4313-b98f-36355ddf4746.png">

#### Old release build
<img width="521" alt="image" src="https://user-images.githubusercontent.com/1931331/186834341-88b65f20-cb05-431b-abb7-1ae788094713.png">

As you can see from 1h 20 min build we are down to 12 min. Quite some time savings. Which still can be slimmed down if we pre-build the containers and re-use them for future builds.  

#skip-changelog

olksdr committed:
feat: Add crossbuild for AARCH64 on X86_64 for relay-cabi lib

These changes basically make sure we cross compile relay-cabi with
proper environment, correct version of libc (the one which is in
manylinux2014_aarch64 - 2.17) and other dependencies.

jan-auer commented on .github/workflows/build_library.yml:
```diff
@@ -31,6 +31,12 @@ jobs:
 
+      - if: matrix.build-arch == 'manylinux2014_aarch64'
+        run: |
+          scripts/docker-aarch64-crossbuild.sh
```
Would it make sense to build the library always in the same container? That way, we could also skip setting up the rust toolchain in the manylinux one.

olksdr replied:
The one reason I added this new step and then reused `manylinux2014_aarch64` for the next step only because python and all the python dependencies are installed inside that container and it just works.  
But I have nothing against  installing correct version of the python and build everything in this one container 👍

jan-auer replied:
That's not what I meant, sorry for the confusion. The two containers are perfectly fine, in fact that will make it easier to align the library build with the binary build once we start cross compiling. 

What I meant is that you can use a docker build for the x86_64 library as well, not just aarch64. That way, there's just a single way to build the library, and that's through docker. That'll make the script / Dockerfile more parametric, but it also allows us to build linux libraries on macOS hosts down the road.

jan-auer commented on scripts/aarch64-libc2_17-cross-build.sh:
```diff
@@ -0,0 +1,28 @@
+dnf --forcearch aarch64 --release 7 install gcc glibc glibc-devel --installroot /usr/aarch64-linux-gnu/sys-root/ -y
+
+# FIX: linker cannot find the libgcc_s, since we have libgcc_s.so.1, so we just link it
+ln -s /usr/aarch64-linux-gnu/sys-root/usr/lib64/libgcc_s.so.1 /usr/aarch64-linux-gnu/sys-root/usr/lib64/libgcc_s.so
```
If you put all of the steps up until here into a dockerfile, we could benefit from caching these steps.

olksdr replied:
Yeah, agree - that's what I mentioned in the description, as a followup for the improvements.  
This PR was just to show the improvement we can gain and if those look are ok.  

I will re-work this script and will use `Dockerfile` for all of this, including installing python dependencies to include building the wheel in one go. And also will add caching for built docker image.

jan-auer replied:
If you look at the current builder image for the main relay binary, you'll notice that it's based on `centos:7` as well. It does not perform a cross build just now, but you might be able to reuse this and just install a cross toolchain in there.

https://github.com/getsentry/relay/blob/984717e894b84570903a2901036212ca559acf47/Dockerfile#L8

iker-barriocanal commented on py/manylinux.sh:
```diff
@@ -7,21 +7,23 @@ yum -y -q install gcc libffi-devel
-curl https://sh.rustup.rs -sSf | sh -s -- -y
-export PATH=~/.cargo/bin:$PATH
+# Install Rust if we have to build the lib inside of this container
+if [ -z $SKIP_RELAY_LIB_BUILD ]; then
```
If I understand correctly, `-z` returns true only if `$SKIP_RELAY_LIB_BUILD` is null (length zero). Should we consider truthy/falsy cases, such as `0` or `1` and `true` or `false`?

olksdr replied:
I don't think we  have to care  here about this. It should be fine to know if it's set or not.  
But also this script can be simplified and the part about the cargo removed - maybe in this PR or as a followup if the proposed changes are accepted

iker-barriocanal replied:
Nah, it's fine.

iker-barriocanal commented on py/setup.py:
```diff
@@ -69,6 +69,10 @@ def delete_scratchpad():
         rust_path = ".."
         scratchpad = None
 
+    # if the lib already built we replace the command
+    if os.environ.get("SKIP_RELAY_LIB_BUILD") is not None:
```
Same here with the truthy/falsy values.

olksdr replied:
same as in https://github.com/getsentry/relay/pull/1438#discussion_r959797823

iker-barriocanal commented on scripts/aarch64-libc2_17-cross-build.sh:
```diff
@@ -0,0 +1,28 @@
+# install our target
+rustup target add aarch64-unknown-linux-gnu
+
+# since the Centos is quite old, and were are on the different architecture, we will disable gpg verification for now
```
I don't really understand why we disable gpg verification. Is the reason to not do too many things for now? It's ok to do that, but I don't get why working on an old version influences the decision.

olksdr replied:
Since the distribution is very old and the script also installs binaries from completely different architecture, the gpg check fails very often. 
Another way would be to introduce some retries/backoffs for fetching keys from the mirrors.

olksdr committed:
ref: Use general docker file to prepare cross compilation

olksdr committed:
ref: Use one image for all the platforms

These commit adds quite few significatn changes:
* bump base image to manylinux2014 (which also bumps libc
  to 2.17)
* use one image, `quay.io/pypa/manylinux2014_x86_64` for all the
  platforms
* we create a builder image per platform with all the dependencies and
  tooling required to build the library
* `using docker/build-push-action@v3` to build images on the fly if
  needed and cache them

This allows to use added Dockefile to cross compile library also locally
for all the supported platforms, with RPM-based linux distribution
(Centos 7 is recomended)

olksdr committed:
Merge branch 'master' into feat/crossbuild-aarch64-on-x86_64

olksdr committed:
ref: update Makefile to use new script to compile library

jan-auer commented on scripts/docker-aarch64-crossbuild.sh:
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
```
This file is not referenced anywhere and can be removed.

jan-auer commented on Makefile:
```diff
@@ -38,8 +38,8 @@ wheel: setup-git setup-venv ## build a wheel of the Python library
 
-wheel-manylinux: setup-git ## build manylinux 2010 compatible wheels
-	@scripts/docker-manylinux.sh
+wheel-manylinux: setup-git ## build manylinux 2014 compatible wheels, you must to have a pre-built image from py/Dockerfile
```
This is unfortunate as we also do not have a utility (e.g. `make` command) to create the builder stage. That makes it hard(er) to reproduce this locally.

Could we move the build of that builder image into the bash script? You can still have the caches in GHA before invoking the script.

olksdr replied:
This target now will build the image and then run the cross build inside that image

jan-auer commented on .github/workflows/build_library.yml:
```diff
@@ -10,31 +10,54 @@ env:
+            TARGET=${{matrix.build-arch.target}}
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+          outputs: type=docker,dest=/tmp/relay-cabi-builder-${{matrix.build-arch.target}}.tar
```
I was not able to find the `outputs` and following load mentioned in the cache examples from this action -- is this indeed required with the `gha` cache type?

olksdr replied:
as discussed I've removed these steps and build the image in the script instead

jan-auer commented on .github/workflows/build_library.yml:
```diff
@@ -10,31 +10,54 @@ env:
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Build and export
```
Can we slightly improve names for these actions? Describe what this builds vs the "Build in Docker" below.

jan-auer commented on .github/workflows/build_library.yml:
```diff
@@ -10,31 +10,54 @@ env:
-          - manylinux2014_aarch64
+          - { image: 'manylinux2014_x86_64', target: 'i686' }
+          - { image: 'manylinux2014_x86_64', target: 'x86_64' }
+          - { image: 'manylinux2014_x86_64', target: 'aarch64' }
```
See comment below in the builder script. We do not actually use the image declared here, and I would suggest we remove `image` from here.

jan-auer commented on .github/workflows/build_library.yml:
```diff
@@ -10,31 +10,54 @@ env:
+      - name: Build and export
+        uses: docker/build-push-action@v3
+        with:
+          context: py/
```
Does this need context? I wasn't able to see why the build container depends on context -- shouldn't it just bundle build dependencies independent of the application code (same as with the `relay-deps` builder of the main binary)?

Please consider that we implicitly bump the MSRV with every stable release, so it might be simpler if we just forgo any caching here. The library is released so rarely that an additional 1-2 minutes does not matter.

jan-auer commented on py/manylinux.sh:
```diff
@@ -7,21 +7,23 @@ yum -y -q install gcc libffi-devel
-curl https://sh.rustup.rs -sSf | sh -s -- -y
-export PATH=~/.cargo/bin:$PATH
+# Install Rust if we have to build the lib inside of this container
+if [ -z $SKIP_RELAY_LIB_BUILD ]; then
```
This is set unconditionally, so you can remove this entire step.

jan-auer commented on scripts/docker-build-lib-linux.sh:
```diff
@@ -0,0 +1,47 @@
+  -e "CARGO_TARGET_${TARGET_LINKER}_UNKNOWN_LINUX_GNU_LINKER" \
+  -e CARGO_BUILD_TARGET \
+  $BUILDER_NAME \
+    bash -c 'cargo build -p relay-cabi --release'
```
```suggestion
  bash -c 'cargo build -p relay-cabi --release'
```

jan-auer commented on scripts/docker-build-lib-linux.sh:
```diff
@@ -0,0 +1,47 @@
+  --rm \
+  -w /work/py \
+  -v "$(pwd):/work" \
+  -e SKIP_RELAY_LIB_BUILD \
```
Since this is set unconditionally, it's more straight forward to assign this directly: 
```suggestion
  -e SKIP_RELAY_LIB_BUILD=1 \
```

jan-auer commented on scripts/docker-build-lib-linux.sh:
```diff
@@ -0,0 +1,47 @@
+
+set -e
+
+IMAGE=${IMAGE:-manylinux2014_x86_64}
```
`IMAGE` is essentially unused and just used to check if the host architecture (which could be obtained via `uname -m`) matches the target architecture. The build actually runs in a `centos:7` based container. 

It is a bit misleading that `manylinux2014` is being used here, but then the actual use is hardcoded below. The image name can for sure be hardcoded this in this script (it used to be originally), but we should get rid of the `IMAGE` env.

olksdr replied:
This `$IMAGE` var is used to provide the name of the image which will be used to build the docker image for cross compilation. The default image is `centos:7`, but we use `manylinux2014_x86_64` since a lot of dependencies already installed inside which simplifies and speeds up preparation of the build environment.  

And also I would like us to be flexible enough with this , to be able to test different base images (e.g. locally) for the build without need to change anything.

jan-auer commented on scripts/docker-build-lib-linux.sh:
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
```
This file has been renamed from or recreated in place of `docker-manylinux.sh`, but it's essentially that. We could rename it back.

olksdr committed:
ref: Move docker build in script and remove unnecessary vars

olksdr committed:
Merge branch 'master' into feat/crossbuild-aarch64-on-x86_64

olksdr committed:
ref: Remove condition which is always will be true

jan-auer commented on scripts/docker-manylinux.sh:
```diff
@@ -1,15 +1,48 @@
+
+
+# Build docker image with all dependencies for cross compilation
+docker build --build-arg=TARGET=${TARGET} --build-arg=IMAGE=${IMAGE} --file py/Dockerfile -t ${BUILDER_NAME} py/
```
```suggestion
docker build --build-arg TARGET=${TARGET} --build-arg IMAGE=${IMAGE} -t ${BUILDER_NAME} py/
```

jan-auer committed:
build(py): Changes from code review

jan-auer committed:
build: Reorder dockerfile

olksdr committed:
ref: always use x86_64 to compare to the requested target

olksdr committed:
ref: remove not relevant changes: always use gpg check in repos

jan-auer committed:
build: Revert to building on amd64

olksdr committed:
feat(ci): Pre-fetch correct gpg for aarch64

In case we have to complie aarch64 on x86_64 we must pre-fetch the
correct gpg key and make sure it will be used.

olksdr committed:
Merge branch 'master' into feat/crossbuild-aarch64-on-x86_64

jan-auer reviewed (approved):
:shipit:
